File: src/net/merayen/elastic/ui/objects/components/PopupParameter1D.java
package net.merayen.elastic.ui.objects.components;

import net.merayen.elastic.ui.UIObject;
import net.merayen.elastic.ui.objects.components.framework.PopupParameter;

/**
 * Presents UI for a single parameter.
 */
public class PopupParameter1D extends UIObject { // TODO rename to PopupParameter
	public interface Handler {
		public void onMove(float value);
		public void onChange(float value);
	}

	private class Window extends UIObject {
		@Override
		protected void onDraw() {
			draw.setColor(150, 150, 150);
			draw.fillRect(-20, 0, 20, popup_height);

			draw.setColor(50, 50, 50);
			draw.rect(-20, 0, 20, popup_height);
		}
	}

	private final PopupParameter box;
	private final UIObject window;
	private final UIObject window_container; // Only for offsetting the popup

	private Handler handler;

	public float popup_height = 200f;
	public final BoxLabel label = new BoxLabel("###");

	public float drag_scale = 1f;

	public PopupParameter1D() {
		window_container = new UIObject();
		window = new Window();
		window_container.add(window);

		box = new PopupParameter(label, window_container);
		box.setHandler(new PopupParameter.Handler() {

			@Override
			public void onMove() {
				box.popup.translation.x = 0; // Constrain X-axis
				if(handler != null)
					//handler.onMove(box.popup.translation.y / popup_height);
					handler.onMove(box.getY());
			}

			@Override
			public void onGrab() {}

			@Override
			public void onDrop() {
				if(handler != null)
					handler.onChange(box.popup.translation.y / popup_height);
			}
		});

		box.popup_height = popup_height;
	}

	@Override
	protected void onInit() {
		add(box);
	}

	@Override
	protected void onUpdate() {
		box.popup_height = popup_height;
		window.translation.y = label.getHeight() / 2;
		box.drag_scale_y = drag_scale;
	}

	public void setHandler(Handler handler) {
		this.handler = handler;
	}

	public UIObject getPane() {
		return box.minified;
	}

	public void setValue(float value) {
		box.setY(value);
	}

	public float getValue() {
		return box.getY();
	}
}

File: src/main/kotlin/net/merayen/elastic/ui/objects/components/PopupParameter1D.kt
package net.merayen.elastic.ui.objects.components

import net.merayen.elastic.ui.Draw
import net.merayen.elastic.ui.FlexibleDimension
import net.merayen.elastic.ui.UIObject
import net.merayen.elastic.ui.objects.components.framework.PopupParameter

/**
 * Presents UI for a single parameter.
 */
class PopupParameter1D : UIObject(), FlexibleDimension {
	override var layoutWidth = 100f
	override var layoutHeight = 20f
	var automaticSizing = true

	interface Handler {
		fun onMove(value: Float)
		fun onChange(value: Float)
		fun onLabel(value: Float): String
	}

	private inner class Window : UIObject() {
		override fun onDraw(draw: Draw) {
			draw.setColor(150, 150, 150)
			draw.fillRect(-20f, 0f, 20f, popup_height)
			draw.setColor(50, 50, 50)
			draw.rect(-20f, 0f, 20f, popup_height)
		}
	}

	private val minified = object : UIObject() {
		override fun onDraw(draw: Draw) {
			draw.setColor(0.2f, 0.2f, 0.2f)
			draw.fillRect(0f, 0f,  layoutWidth + margin * 2, 12f + margin * 2)

			draw.setStroke(1f)
			draw.setColor(0.7f, 0.7f, 0.7f)
			draw.rect(0f, 0f, layoutWidth + margin * 2, 12f + margin * 2)
		}
	}

	private val box: PopupParameter
	private val window: UIObject
	private val window_container: UIObject // Only for offsetting the popup
	var handler: Handler? = null
	var popup_height = 200f

	private val margin = 4f // ???

	@JvmField
	val label = Label()

	@JvmField
	var drag_scale = 1f

	override fun onInit() {
		add(box)
		minified.add(label)
	}

	override fun onUpdate() {
		box.popup_height = popup_height
		window.translation.y = label.getHeight() / 2
		box.drag_scale_y = drag_scale
		label.translation.x = margin
		label.translation.y = margin
		getWidth()

		if (automaticSizing)
			layoutWidth = label.labelWidth
	}

	val pane: UIObject
		get() = box.minified

	var value: Float
		get() = box.y
		set(value) {
			box.y = value
			label.text = handler?.onLabel(value) ?: ""
		}

	init {
		window_container = UIObject()
		window = Window()
		window_container.add(window)

		box = PopupParameter(minified, window_container)
		box.setHandler(object : PopupParameter.Handler {
			override fun onMove() {
				box.popup.translation.x = 0f // Constrain X-axis
				handler?.onMove(box.y)
				label.text = handler?.onLabel(box.y) ?: ""
			}

			override fun onGrab() {}
			override fun onDrop() {
				handler?.onChange(box.popup.translation.y / popup_height)
				label.text = handler?.onLabel(box.y) ?: ""
			}
		})
		box.popup_height = popup_height
	}
}